unic/utils/src/macros.rs

```diff
@@ -0,0 +1,341 @@
+///
+/// # Syntax (Enumerated Property)
+///
+/// ```
+/// # #[macro_use] extern crate unic_utils;
+/// # fn main() {}
+/// char_property! {
+///     /// Zero or more attributes
+///     pub enum PropertyName {
+///         /// Exactly one attribute
+///         RustName {
+///             abbr => AbbrName,
+///             long => Long_Name,
+///             display => "&'static str that is a nicer presentation of the name",
+///         }
+///
+///         /// All annotations on the variant are optional*
+///         Variant2 {
+///             abbr => V2, // *abbr is required for Enumerated Properties
+///         }
+///     }
+///
+///     /// Zero or more attributes
+///     pub mod abbr_names;
+///
+///     /// Zero or more attributes
+///     pub mod long_names;
+/// }
+///
+/// // You must impl (Partial/Complete)CharProperty manually.
+/// # impl unic_utils::char_property::PartialCharProperty for PropertyName {
+/// #     fn of(_: char) -> Option<Self> { None }
+/// # }
+/// ```
+///
+/// # Effect
+///
+/// - Implements `CharProperty` with the `abbr` and `long` presented in the appropriate method
+/// - Implements `FromStr` accepting any of the rust, abbr, or long names
+/// - Implements `Display` using the given string, falling back when not provided on
+///   the long name, the short name, and the rust name, in that order
+/// - Populates the module `abbr_names` with `pub use` bindings of variants to their abbr names
+/// - Populates the module `long_names` with `pub use` bindings of variants to their long names
+/// - Maintains all documentation comments and other `#[attributes]` as would be expected
+///   (with some caveats, listed below)
+///
+/// # Limitations
+///
+/// Due to [rust-lang/rust/#24189](https://github.com/rust-lang/rust/issues/24189), (fixed in
+/// [rust-lang/rust/#42913](https://github.com/rust-lang/rust/pull/42913), landing in 1.20),
+/// exactly one attribute line must be used on each variant. On 1.20 or higher, one or more may
+/// be used, and the restriction can be relaxed back the intended zero or more by replacing
+/// `$(#[$variant_meta:meta])+` with `$(#[$variant_meta:meta])*`, and
+/// `$(#[$variant_meta])+` with `$(#[$variant_meta])*`, and
+/// `$(#[$ident_meta:meta])+` with `$(#[$ident_meta:meta])*` and
+/// `$(#[$ident_meta])+` with `$(#[$ident_meta])*`, and
+/// `$(#[$rest_meta:meta])+` with `$(#[$rest_meta:meta])*`, and
+/// `$(#[$rest_meta])+` with `$(#[$rest_meta])*`, and
+/// `$(#[$queue_meta:meta])+` with `$(#[$queue_meta:meta])*`, and
+/// `$(#[$queue_meta])+` with `$(#[$queue_meta])*`
+// TODO: Once adopting 1.20, fix the macro to work with zero attributes on variants (see above)
+#[macro_export]
+macro_rules! char_property {
+    (
+        $(#[$name_meta:meta])* pub enum $name:ident {
+            $( $(#[$variant_meta:meta])+ $variant:ident $tt:tt )*
+        }
+
+        $(#[$abbr_names_meta:meta])* pub mod $abbr_names:ident;
+        $(#[$long_names_meta:meta])* pub mod $long_names:ident;
```

Comment:
I think we need to do something about these two modules. The thing it that, in the pattern, there's only one `enum`, which is clear what it is. The two `pub mod` lines, for anyone reading the code, or even developing, it's not clear at all that the first will be used for abbr-names, and the second for long-names. IMHO, we should enforce that in the patter.

For example, something like this:
```

char_property! {
    pub enum MyProperty {
        ...
    }

    pub mod my_property_abbr_names for abbr_names;
    pub mod my_property_long_names for long_names;
}
```

Rational: `for` is a reserved keyword and gets highlighted accordingly. `abbr_names` and `long_names` go well with the names being impl for the enum, `abbr_name`/`long_name`, so it's easy to remember and easy to read/understand.